app/models/language/tag.rb

```diff
@@ -0,0 +1,179 @@
+  validates_with Language::Validators::SubtagsTypeValidator
+  validates_with Language::Validators::ExtLangValidator
+
+  def first_subtag_of_type(type)
+    TagValue.first_subtag_of_type(subtags, type)
+  end
+
+  def suppressed_script
+    first_subtag_of_type('language').suppress_script
+  end
+
+  def script
+    script = first_subtag_of_type('script')
+    script unless script == suppressed_script
+  end
+
+  def region
+    first_subtag_of_type('region')
+  end
+
+  def extlang
+    first_subtag_of_type('extlang')
+  end
+
+  def tag_value(use_preferred = false)
+    TagValue.for_subtags(subtags, use_preferred)
+  rescue SubtagError => ex
+    self.errors.add :subtags, ex.message
+  end
+
+  def set_tag
+    return if tag_type == 'grandfathered'
+    self.tag = tag_value
+  end
+
+  def set_preferred_value
+    return if preferred_value || tag_type.eql?('grandfathered')
+    language = first_subtag_of_type('language')
+    script = first_subtag_of_type('script')
+    extlang = first_subtag_of_type('extlang')
+    has_preferred = language.preferred_value
+    has_preferred ||= script && (language.suppress_script == script)
+    has_preferred ||= extlang&.preferred_value
+    return unless has_preferred
+    self.preferred_value = ::Language::Tag.for(TagValue.for_subtags(subtags, true))
+  end
+
+  def use_value
+    preferred_value || self
+  end
+
+  # accepts a BCP-47 tag and returns the corresponding canonicalized or preferred tag
+  # raises SubtagError if tags are invalid
+  def self.for(tag_value, use_preferred = true)
+    tag = ::Language::Tag.find_by(tag: tag_value) || compose_tag_for(tag_value)
+    (use_preferred && tag.preferred_value) || tag
+  rescue ActiveRecord::RecordInvalid => ex
+    raise SubtagError, ex.message.sub('Validation failed: ', '')
+  end
+
+  def self.compose_tag_for(tag_value)
+    subtags = subtags_for(tag_value.split('-'))
+    tag_value = TagValue.for_subtags(subtags)
+    candidate = ::Language::Tag.find_by(tag: tag_value) || ::Language::Tag.new(added: Time.new.getlocal, subtags: subtags)
+    candidate.set_tag
+    tag = ::Language::Tag.find_by(tag: candidate.tag) || candidate
+    tag.save! unless tag.persisted?
+    tag
+  end
+
+  def self.subtags_for(subtag_values)
+    subtags = ::Language::Subtag.where(subtag: subtag_values).to_a
+    extlang = subtags.detect { |st| st.subtag_type.eql?('extlang') }
+    # if a macrolanguage-extlang combination has a preferred form,
+    # there may be a duplicate language tag for the extlang
+    if extlang
+      other_language = subtags.detect { |st| !st.subtag.eql?(extlang.subtag) && st.subtag_type.eql?('language') }
+      duplicate = subtags.detect { |st| st.subtag.eql?(extlang.subtag) && st.subtag_type.eql?('language') }
+      if duplicate
+        if other_language
+          subtags.delete(duplicate)
+        else
+          subtags.delete(extlang)
+        end
+      end
+    end
+    subtags
+  end
+  class SubtagError < RuntimeError; end
```

Comment:
I often subclass StandardError for custom error classes, but maybe I should be using RuntimeError instead like you've done here.

Reply:
Rubocop's suggestion - it is just one subclass away!